File: src/WordSearch.java
public class WordSearch{
	public boolean exist(char[][] board, String word){
		int ROW = board.length;
		int COL = board[0].length;
		if (word.length() == 0)		return true;
		else if (word.length() > ROW * COL)		return false;
		boolean[][] visited = new boolean[ROW][COL];
		for (int indexR = 0; indexR < ROW; indexR++){
			for (int indexC = 0; indexC < COL; indexC++){
				if (board[indexR][indexC] == word.charAt(0)){
					visited[indexR][indexC] = true;
					if (search(board, indexR, indexC, visited, word.substring(1)))	return true;
					visited[indexR][indexC] = false;
				}
			}
		}
		return false;
	}
		
	private boolean search(char[][] board, int row, int col, boolean[][] visited, String word){
		if (word.length() == 0)	return true;
		int[][] pos = {{row + 1, col}, {row - 1, col}, {row, col + 1}, {row, col - 1}};
		for (int index = 0; index < 4; index++){
			int newRow = pos[index][0];
			int newCol = pos[index][1];
 			if (newRow < board.length && newRow >= 0 
							&& newCol < board[0].length && newCol >= 0 ){
				if (!visited[newRow][newCol] && board[newRow][newCol] == word.charAt(0)){
					visited[newRow][newCol] = true;
					if (search(board, newRow, newCol, visited, word.substring(1))) return true;
					visited[newRow][newCol] = false;
				}
			}
		}
		return false;
	}
	public static void main(String[] args){
		char[][] board= {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
		String word = "ABCCED";
		System.out.println(new WordSearch().exist(board, word));
	}
}
				

File: python/word_search.py
#! /usr/bin/python

'''
 Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell,
where "adjacent" cells are those horizontally or vertically neighboring.
 The same letter cell may not be used more than once.

For example,
Given board =

[
  ["ABCE"],
  ["SFCS"],
  ["ADEE"]
]

word = "ABCCED", -> returns true,
word = "SEE", -> returns true,
word = "ABCB", -> returns false.
'''

class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        row    = len(board)
        column = len(board[0])
        size   = len(word)
        visited = [ [False] * column for index in range(row) ]
        for index_row in range(row):
            for index_column in range(column):
                if self.search_recursive(board, visited, index_row, index_column, word):
                    return True
        return False

    def search_recursive(self, board, visited, index_row, index_column, word):
        if len(word) == 0:
            return True
        if index_row < 0 or index_row >= len(board) or index_column < 0 or index_column >= len(board[0]):
            return False
        if visited[index_row][index_column] or board[index_row][index_column] != word[0]:
            return False
        visited[index_row][index_column] = True
        res =  self.search_recursive(board, visited, index_row-1, index_column, word[1:]) \
            or self.search_recursive(board, visited, index_row+1, index_column, word[1:]) \
            or self.search_recursive(board, visited, index_row, index_column-1, word[1:]) \
            or self.search_recursive(board, visited, index_row, index_column+1, word[1:])
        visited[index_row][index_column] = False
        return res


if __name__ == '__main__':
    solution = Solution()
    board = ['ABCE','SFCS','ADEE']
    print solution.exist(board, 'ABCCED')
    print solution.exist(board, 'SEE')
    print solution.exist(board, 'ABA')
    board = ['A']
    print solution.exist(board, 'AB')
    print solution.exist(["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaaaaaaaaaaaaab"], "baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
